File: Problems/Algorithms/82. Remove Duplicates from Sorted List II/RemoveDuplicatesII.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun deleteDuplicates(head: ListNode?): ListNode? {
        var dummy: ListNode? = ListNode(0)
        var curr: ListNode? = head
        var prev: ListNode? = dummy
        
        while (curr != null) {
            if (curr?.`val` == curr?.next?.`val`) {
                while (curr?.`val` == curr?.next?.`val`) {
                    curr = curr?.next
                }
            } else {
                prev?.next = curr
                prev = curr
            }
            curr = curr?.next
        }
        prev?.next = curr
        
        return dummy?.next
    }
}

File: Problems/Algorithms/82. Remove Duplicates from Sorted List II/remove_duplicates_II.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func deleteDuplicates(head *ListNode) *ListNode {
    dummy := &ListNode{0, nil}
    curr := head
    prev := dummy

    for curr != nil && curr.Next != nil {
        if curr.Val == curr.Next.Val {
            for curr != nil && curr.Next != nil && curr.Val == curr.Next.Val {
                curr = curr.Next
            }
        } else {
            prev.Next = curr
            prev = curr
        }
        curr = curr.Next
    }
    prev.Next = curr

    return dummy.Next
}
